### DISCUSSION TITLE
Why is shallow always false when using startTransition

### DISCUSSION THREAD

**Discussion by zfm-dennisdisselhoff** [47ng/nuqs#612] [Q&A] [answered]
I came up with the following problem: 

We are using react-query in one of our next (app directory) projects. For loading data from a server, we implemented the useSuspenseQuery hook. Since version [v5.24.2](https://github.com/TanStack/query/releases/tag/v5.24.2), we experienced a loading screen when the underlying querykey of the current query is changing. This happens because the result of useSuspenseQuery doesn't show the previousData anymore, instead the app renders the nearest suspense boundary. The solution from the react-query docs suggests to use startTransitions to avoid this behavior. In next-usequerystate it is possible to give the useQueryState hook an startTransition function. The Problem is, that this will set the shallow option to false. This results in a server roundtrip and our app is showing a loading screen. Is there any reason the startTransition property is not allowed to be used with the enabled shallow property?

**Comment by franky47** [comment, marked as answer]
There are two things that happen when you call the state updater function of a `useQueryState(s)` hook:

1. The internal state is updated immediately, and synced across hooks with the same key.
2. At a later time, the URL is updated via a navigation event. What actually happens depends on your setup and configuration:
  a. App router + shallow update: History API (`history.replaceState`)
  b. App router + non-shallow update: Router call (updates the server)
  c. Pages router: Router call with the shallow option passed along

Why is point n°2 happening later, you may ask? Because browsers rate-limit URL updates, some aggressively (looking at you, Safari). And even so, some form of rate-limiting/thottling is required on non-shallow updates to avoid spamming the server.

This decoupling allows `nuqs` to behave like a drop-in replacement for `useState`, with eventual URL consistency.

Now, for the transition part, as you can see, there are two events we may want to wrap into transitions. It might make sense to do the internal state update in a transition if the state value is used for heavy synchronous computation that may slow the event loop down. In this case, it's not about getting progress information back, but rather performance optimisation by letting React do the state update when it has enough bandwidth.

The `startTransition` argument you pass to `nuqs` is used for point n°2, when the navigation event happens. Focusing on the app router (I don't know if transitions are supported on the pages router), transitions on navigation events are only supported on router API calls, not on the history API (shallow updates). This is why passing a `startTransition` option opts-in to updating the server: so you can actually measure progress with `useTransition`'s `isLoading` state.

Now, if you need to keep shallow routing to avoid hitting the server, and want to avoid suspending on state updates that may trigger a TanStack query update, you can wrap the `useQueryState` update function in a `startTransition`, which acts like deferring a `useState`:

```tsx
import { startTransition, useCallback } from 'react'
import { useQueryState } from 'nuqs'

const useQueryStateWithTransition() {
  const [query, setQuery]  = useQueryState('q')
  const update = useCallback((value: string | null) => {
    startTransition(() => {
      setQuery(value)
    })
  }, [setQuery])
  return [query, update] as const
}
```

**Reply by zfm-dennisdisselhoff** [reply]
Thank you for your detailed answer! This helped a lot. In our case, the useTransition hook caused further errors, but we could use the useDeferred hook to achieve our goals.
